### ISSUE TITLE
dnn miss many tensorflow layer type

### ISSUE THREAD

**Issue by vonchenplus** [opencv/opencv#14683]
I'm using the readNetFromTensorflow to import a tensorflow pb frozen model. However, I get an error during forward:
Can't create layer "moments/StopGradient" of type "StopGradient"

and i found dnn miss many other tensorflow layer type,for example:
SquaredDifference
Power,
Pow,
RealDiv,
TanH

so, opencv have any plan to add these layers? Or how could I remove this as this is not used in the network any longer after training.

thanks.

**Comment by dkurt** [comment]
`StopGradient` is not a deployment node - turn the model into testing phase.
Power, TanH - supported

Pow, SquaredDifference, RealDiv most likely are part of Batch Normalization subgraphs (we can fuse ift automatically)

Please provide reproducer if it's a bug report.

**Comment by vonchenplus** [comment]
hello dkurt, thanks for your reply.
so, how can i fixed this problem?
I don't quite understand how opencv handles this situation.
I upload a pb file at google drive:
https://drive.google.com/open?id=1KGTYKEhH0UZyQtMObAE5WVW8MZ7DFfoP
Please help me to see this model when running with opencv.
thank you very much.

**Comment by dkurt** [comment]
@vonchenplus, the model is in training phase. You need to switch it to testing mode.

**Comment by dkurt** [comment]
@vonchenplus, Thanks for report! Please try the changes from https://github.com/opencv/opencv/pull/14747. I tested your model in the following way:

```python
import cv2 as cv
import numpy as np

import tensorflow as tf

np.random.seed(324)

cvNet = cv.dnn.readNet('la_muse.pb')
inp = np.random.standard_normal([1, 3, 256, 256]).astype(np.float32)
cvNet.setInput(inp)
cvOut = cvNet.forward()

print(cvOut.shape)

graph = 'la_muse.pb'
with tf.gfile.FastGFile(graph) as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Session() as sess:
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')
    tfOut = sess.run(sess.graph.get_tensor_by_name('add_37:0'),
                     feed_dict={'ContentImage:0': inp.transpose(0, 2, 3, 1)})

print(np.max(np.abs(tfOut.transpose(0, 3, 1, 2) - cvOut)))
print(np.min(cvOut), np.max(cvOut))
```

It shows that maximal absolute difference is about `0.004`. Output values range is `(-22.5, 277.5)` so I think that difference is OK (about `1.617432e-05` of entire range).

**Comment by vonchenplus** [comment]
@dkurt, Thanks for report! 
This path work for me, thanks!

**Comment by vonchenplus** [comment]
@dkurt , I set preferable target to DNN_TARGET_OPENCL, it will been blocked at ocl.cpp:5056
`u->data = (uchar*)clEnqueueMapBuffer(q, (cl_mem)u->handle, CL_TRUE,
                                                         (CL_MAP_READ | CL_MAP_WRITE),
                                                         0, u->size, 0, 0, 0, &retval);`

follow the call stack:
`igdrclneo64.dll!00007ffbbd784302()	Unknown
igdrclneo64.dll!00007ffbbd7ee52e()	Unknown
igdrclneo64.dll!00007ffbbd7b5e91()	Unknown
igdrclneo64.dll!00007ffbbd7ea120()	Unknown
igdrclneo64.dll!00007ffbbd869584()	Unknown
igdrclneo64.dll!00007ffbbd86b08a()	Unknown
igdrclneo64.dll!00007ffbbd772836()	Unknown
OpenCL.DLL!00007ffbddbd361f()	Unknown
cv::ocl::OpenCLAllocator::map(cv::UMatData * u, cv::AccessFlag accessFlags) Line 5056	C++
cv::UMat::getMat(cv::AccessFlag accessFlags) Line 825	C++
cv::UMat::setTo(const cv::debug_build_guard::_InputArray & _value, const cv::debug_build_guard::_InputArray & _mask) Line 1095	C++
cv::UMat::operator=(const cv::Scalar_<double> & s) Line 1102	C++
cv::UMat::UMat(int _dims, const int * _sz, int _type, const cv::Scalar_<double> & _s, cv::UMatUsageFlags _usageFlags) Line 3635	C++
cv::UMat::zeros(int ndims, const int * sz, int type) Line 1207	C++
`

And Some other window, for example:chrome, may be appears black screen.
I think this should be a resource leak.

**Comment by alalek** [comment]
> And Some other window, for example:chrome, may be appears black screen.

This can be caused by huge computation tasks - what is "CPU" inference time for your model?
GPU is busy, so it can't process other GPU tasks (drawing chrome windows).
Try to recompile OpenCV with [OpenCL subsystem debugging](https://github.com/opencv/opencv/blob/3.4.6/modules/core/src/ocl.cpp#L68-L69) (set mentioned defines to "1")

---

Please provide information about used OpenCL and versions:
- output from `clinfo` or `opencv_version --opencl`

**Comment by vonchenplus** [comment]
My OpenCL Version is 1.2,
Here is Opencv debug log:
https://drive.google.com/open?id=1yDjvW9fBwkNNt7lllO2AQLqtiDAbAjWj

**Comment by alalek** [comment]
@vonchenplus Did you try to run OpenCV DNN tests? Do they freeze?

BTW, Your driver version is 23.20.16.4973. It has been released ~1 year ago.
Up-to-date Intel Graphics driver can be found [here](https://downloadcenter.intel.com/product/80939/Graphics-Drivers).

**Comment by vonchenplus** [comment]
@alalek, I am update Intel Graphics driver yesterday and it's working properly now.

But, I found the Opencl mode slower than cpu mode, Is this normal?

**Comment by alalek** [comment]
Great!

Only subset of convolution modes has been properly optimized/[configured](https://github.com/opencv/opencv/blob/3.4.6/modules/dnn/src/ocl4dnn/include/default_kernel_config.hpp) for OpenCL computations (they cover most cases of public models from OpenCV tests). All other cases are go via general "naive" implementation (and it is sometimes slower than CPU).

You can try to enable OpenCL tuning via environment variables (this mode has slow initialization at first time):
- `OPENCV_OCL4DNN_CONFIG_PATH=<some_existed_emoty_directory>`
- remove contents of that directory to rerun tuning phase again or use `OPENCV_OCL4DNN_FORCE_AUTO_TUNING=1`

Another important note - all layers should be run on OpenCL - switching between OpenCL/CPU increases computation time too.
You can check what is going with your model in VTune ([it is free](https://software.intel.com/en-us/vtune/choose-download)) - select [GPU hotspots analysis](https://software.intel.com/en-us/vtune-amplifier-help-gpu-opencl-application-analysis-view) (ensure that your first "tuning" run is completed before VTune usage).

**Comment by vonchenplus** [comment]
hello @alalek , OCL4DNN initialization too slow, i give up. 

I think this issue has no problem for me, need me to close the issue?

**Comment by vonchenplus** [comment]
hello @dkurt , I test my new model and i found dnn unsupport these operation:
MirrorPad,
MaxPoolGrad(Very similar to MaxUnpoolLayer)

**Comment by vonchenplus** [comment]
hello @dkurt, I Pull request to support MirrorPad at [14832](https://github.com/opencv/opencv/pull/14832), please help me review it.

**Comment by dkurt** [comment]
@vonchenplus, Can we close this issue due merged https://github.com/opencv/opencv/pull/14860?
